### PR TITLE
Feature/beta parameter

### PR DESCRIPTION
**Pull request by PietropaoloFrisoni**
Adding computation of the beta parameter for the portfolio. In the Capital Asset Pricing Model (CAPM), the beta parameter defines how risky the portfolio is compared to the market. 

Important: I could not run the old tests due to the usual _quandl_ error (although I obtained my API key). I still don't know what's happening, but I'm working on it. The old tests should be run as a sanity check before merging the pull request.

However, the additional feature is written so that all the previous codes and tests should still work: suppose the argument "market_index" is not explicitly passed to the "build_portfolio" function. In that case, the Market class instance is not created, and the beta parameter of the portfolio is not computed or printed in properties.

@fmilthaler if there's something you would like me to change, please tell me.

All the best

### CODE REVIEW

**Comment by fmilthaler** [comment]
Thank you so much for your past contributions and this big one! It's a big help.

Don't worry about the quandl error, you can ignore those for the time being.

I'll have a closer look at it over the next few days. Once this is merged, I'll create a new release for FinQuant.

**Comment by PietropaoloFrisoni** [comment]
Excellent, thank you! Please take your time and let me know if I should change something.

Best

File: finquant/portfolio.py [hunk 1]
```diff
@@ -818,22 +865,32 @@ def _build_portfolio_from_api(
          - ``quandl`` (Python package/API to `Quandl`)
          - ``yfinance`` (Python package formerly known as ``fix-yahoo-finance``)
```
Comment (fmilthaler) [inline finquant/portfolio.py:867]:
```suggestion
```

File: finquant/portfolio.py [hunk 2]
```diff
@@ -987,6 +1070,9 @@ def build_portfolio(**kwargs):
...
```
Comment (fmilthaler) [inline finquant/portfolio.py:1072]:
```suggestion
```

File: finquant/portfolio.py [hunk 3]
```diff
@@ -1055,6 +1142,7 @@ def build_portfolio(**kwargs):
         "start_date",
         "end_date",
         "data_api",
+        "market_index",
```
Comment (fmilthaler) [inline finquant/portfolio.py:1145]: add market_index to the comment above as well: # 1. pf_allocation, names, start_date, end_date, data_api, market_index

File: finquant/portfolio.py [hunk 4]
```diff
@@ -987,6 +1070,9 @@ def build_portfolio(**kwargs):
          - ``quandl`` (Python package/API to `Quandl`)
          - ``yfinance`` (Python package formerly known as ``fix-yahoo-finance``)
 
+     :market_index: (optional) ``string`` (default: ``None``) which determines the
```
Comment (fmilthaler) [inline finquant/portfolio.py:1073]: add market_index to the docstring (for the documentation) below as well:
```
- ``names``, ``pf_allocation`` (optional), ``start_date`` (optional), ``end_date`` (optional), ``data_api`` (optional), ``market_index``
```

**Comment by fmilthaler** [comment]
Great addition @PietropaoloFrisoni. Mainly just some minor changes, e.g. adding comment or removing empty line.

With the introduction of the class `Market`, which has a huge overlap with `Stock`, we should rather come up with a class `Asset`, define common methods in there, and then have `Stock` and `Market` inherit from `Asset`. That would be good. However, I'm fine to have that change done in a separate PR. Your decision, if you want to do it later, then please open an issue for that.

One additional change would be great for this PR: Could you add this as an example? You could make some modifications to `example/Example-Build-Portfolio-from-web.py`. Currently, that example uses quandl to retrieve data. Since quandl is basically dead, feel free to also change that example to use `yfinance` instead, but leave all the other mentions of quandl for now. Removing quandl properly should be done in another PR (some day in the future) ;)

**Review by fmilthaler — CHANGES_REQUESTED** [review]

File: finquant/market.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
+"""This module provides a public class ``Market`` that holds and calculates quantities of a market index"""
+
+import numpy as np
+import pandas as pd
+
+from finquant.returns import historical_mean_return, daily_returns
+
+
+class Market(object):
```
[the remark this answers is not in the document]
Comment (PietropaoloFrisoni) [reply finquant/market.py:9]: Yes, it looks quite reasonable. I agree that the most straightforward approach is creating an issue to work on in the future. I'll make it myself after this PR has been merged.

**Comment by PietropaoloFrisoni** [comment]
@fmilthaler Hi, I'm following your suggestions and updating with commits. However, the checks are currently failing due to _quandl_. At line 38 of the build (3.11) test, I see that: 

_quandl.errors.quandl_error.LimitExceededError: (Status 429) (Quandl Error QELx01) You have exceeded the anonymous user limit of 50 calls per day. To make more calls today, please register for a free Nasdaq Data Link account and then include your API key with your requests._ 

I may be missing something, but it looks like this is the source of the error. Any idea how to solve this?

File: finquant/portfolio.py [hunk 6]
```diff
@@ -1021,7 +1021,7 @@ def _build_portfolio_from_df(
         # extract only "Adjusted Close" price column from market data
         market_data = _get_index_adj_clos_pr(market_data)
         # set market index of portfolio
-        pf.set_market_index(Market(data=market_data))
+        pf.market_index = Market(data=market_data)
```
Comment (fmilthaler) [inline finquant/portfolio.py:1024]: 👍

**Comment by fmilthaler** [comment]
> @fmilthaler Great, tests work now. I'll work on the `example/Example-Build-Portfolio-from-web.py` as soon as possible (tomorrow at the latest)

Thank you, that would be most helpful for the users (including myself) to see how to use it. :)

**Comment by PietropaoloFrisoni** [comment]
Hi @fmilthaler, I updated the example as well. Please tell me if there's something you would like to change.

Best

**Comment by fmilthaler** [comment]
@PietropaoloFrisoni Amazing work, I really appreciate all the work you put in here!!! You are a star. Just one minor fix I flagged. Once that is in, this can be merged (and yet another release is going to happen) :)

**Comment by PietropaoloFrisoni** [comment]
That's way too kind. Thank you so much. Maybe I'm missing something, as I cannot see the flagged minor issue. Could you please point that out (again)?

**Comment by fmilthaler** [comment]
> Could you please point that out (again)?
That is weird. I commented on it again and tagged you. In case it still does not show up for you:
file: [example/Example-Build-Portfolio-from-web.py, line 8](https://github.com/fmilthaler/FinQuant/pull/84/files#diff-64c0d3234e22320e1a0140aa04850b51e9ec3dd31f97fda9d7dcb2c5a5170f2fR8). Change that line to:
```
# This example only focuses on how to use `build_portfolio()` to get an instance of `Portfolio` by providing a few items of information that is passed on to `quandl`/`yfinance`. For a more exhaustive description of this package and example, please try `Example-Analysis` and `Example-Optimisation`.
```
Only change in that (long) line is the addition of "items of" behind "providing a few". :)

**Comment by PietropaoloFrisoni** [comment]
Done. Thanks for pointing that out!

**Review by fmilthaler — APPROVED** [review]

**Review by fmilthaler — APPROVED** [review]

**Pull request merged**